File: src/main/java/net/gtaun/shoebill/object/PlayerMapIcon.java
package net.gtaun.shoebill.object;

import net.gtaun.shoebill.constant.MapIconStyle;
import net.gtaun.shoebill.data.Color;
import net.gtaun.shoebill.data.Vector3D;

public interface PlayerMapIcon extends PlayerRelated
{
	public interface MapIcon extends Destroyable, PlayerRelated
	{
		int getId();

		void update(float x, float y, float z, int markerType, Color color, MapIconStyle style);
		void update(Vector3D pos, int markerType, Color color, MapIconStyle style);
	}
	
	MapIcon getIcon(int iconid);
	void setIcon(int iconid, MapIcon icon);
	MapIcon createIcon(float x, float y, float z, int markerType, Color color, MapIconStyle style);
	MapIcon createIcon(Vector3D pos, int markerType, Color color, MapIconStyle style);
	MapIcon createIcon(float x, float y, float z, int markerType, Color color, MapIconStyle style, int iconId);
	MapIcon createIcon(Vector3D pos, int markerType, Color color, MapIconStyle style, int iconId);
	MapIcon createIcon(int iconId);
	MapIcon createIcon();
}

File: src/main/java/net/gtaun/shoebill/entities/PlayerMapIcon.kt
package net.gtaun.shoebill.entities

import net.gtaun.shoebill.constant.MapIconStyle
import net.gtaun.shoebill.data.Color
import net.gtaun.shoebill.data.Vector3D

/**
 * This class provides functions and methods to handle [Player] map icons.
 *
 * @author MK124
 * @author Marvin Haschker
 */
interface PlayerMapIcon : PlayerRelated {
    interface MapIcon : Destroyable, PlayerRelated {
        val id: Int

        fun update(x: Float, y: Float, z: Float, markerType: Int, color: Color, style: MapIconStyle)
        fun update(pos: Vector3D, markerType: Int, color: Color, style: MapIconStyle)
    }

    fun getIcon(iconid: Int): MapIcon?
    fun setIcon(iconid: Int, icon: MapIcon)
    fun createIcon(x: Float, y: Float, z: Float, markerType: Int, color: Color, style: MapIconStyle): MapIcon
    fun createIcon(pos: Vector3D, markerType: Int, color: Color, style: MapIconStyle): MapIcon
    fun createIcon(x: Float, y: Float, z: Float, markerType: Int, color: Color, style: MapIconStyle, iconId: Int): MapIcon
    fun createIcon(pos: Vector3D, markerType: Int, color: Color, style: MapIconStyle, iconId: Int): MapIcon
    fun createIcon(iconId: Int): MapIcon
    fun createIcon(): MapIcon
}
